### ISSUE TITLE
InvalidFrameBufferOperation when closing example code

### ISSUE THREAD

**Issue by Hittherhod** [amethyst/amethyst#515]
When running example code, I get

```
thread 'main' panicked at 'Error InvalidFramebufferOperation executing command: Clear(Some(Float([0.00196, 0.23726, 0.21765, 1])), None, None)'
```

## Meta

```
Initializing Amethyst...
Version: v0.5.1-394-gaf76188
Platform: x86_64-apple-darwin
Amethyst git commit: af76188a7628c87f306e380df1dc80ec5cfadcbb
Rustc version: 1.22.1 Stable
Rustc git commit: 05e2e1c41414e8fc73d0f267ea8dab1a3eeeaa99
```

## Stack trace

```
stack backtrace:
   0: std::sys::imp::backtrace::tracing::imp::unwind_backtrace
             at src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
   1: std::sys_common::backtrace::_print
             at src/libstd/sys_common/backtrace.rs:69
   2: std::panicking::default_hook::{{closure}}
             at src/libstd/sys_common/backtrace.rs:58
             at src/libstd/panicking.rs:381
   3: std::panicking::default_hook
             at src/libstd/panicking.rs:397
   4: std::panicking::begin_panic
             at src/libstd/panicking.rs:577
   5: std::panicking::begin_panic
             at src/libstd/panicking.rs:538
   6: std::panicking::try::do_call
             at src/libstd/panicking.rs:522
   7: gfx_device_gl::Device::process
             at /Users/rhodysn1/.cargo/registry/src/github.com-1ecc6299db9ec823/gfx_device_gl-0.14.6/src/lib.rs:777
   8: gfx_device_gl::Device::no_fence_submit
             at /Users/rhodysn1/.cargo/registry/src/github.com-1ecc6299db9ec823/gfx_device_gl-0.14.6/src/lib.rs:784
   9: gfx_device_gl::Device::track_mapped_gpu_access
             at /Users/rhodysn1/.cargo/registry/src/github.com-1ecc6299db9ec823/gfx_device_gl-0.14.6/src/lib.rs:906
  10: <gfx::encoder::Encoder<R, C>>::flush_no_reset
             at /Users/rhodysn1/.cargo/registry/src/github.com-1ecc6299db9ec823/gfx-0.16.2/src/encoder.rs:194
  11: <gfx::encoder::Encoder<R, C>>::flush
             at /Users/rhodysn1/.cargo/registry/src/github.com-1ecc6299db9ec823/gfx-0.16.2/src/encoder.rs:185
  12: amethyst_renderer::renderer::Renderer::draw
             at ./amethyst_renderer/src/renderer.rs:84
  13: <amethyst_renderer::system::RenderSystem<P>>::render
             at ./amethyst_renderer/src/system.rs:124
  14: <amethyst_renderer::system::RenderSystem<P> as shred::system::RunNow<'a>>::run_now
             at ./amethyst_renderer/src/system.rs:159
  15: amethyst::app::Application::advance_frame
             at src/app.rs:202
  16: <shred::cell::RefMut<'a, T> as core::ops::deref::DerefMut>::deref_mut
             at src/app.rs:129
  17: window::run
             at examples/window/main.rs:49
  18: window::main
             at examples/window/main.rs:55
  19: panic_unwind::dwarf::eh::read_encoded_pointer
             at src/libpanic_unwind/lib.rs:99
  20: <std::rand::reader::ReaderRng<R> as rand::Rng>::fill_bytes
             at src/libstd/panicking.rs:459
             at src/libstd/panic.rs:361
             at src/libstd/rt.rs:59
  21: window::main
```

when I try to close with the X button.
I do not get this when I close by hitting the Escape key.

I tried to fix this by adding a `Trans::Quit` in the `handle_event` function upon `WindowEvent::Closed`, but that doesn't fix this issue either. 

Does anyone know why this is happening? It's not limited to just one example.

**Comment by Hittherhod** [comment]
Sorry for the bad stack trace formatting

**Comment by torkleyy** [comment]
This is the relevant line on GitHub:

https://github.com/gfx-rs/gfx/blob/pre-ll/src/backend/gl/src/lib.rs#L778

It's located in `gfx_gl_device`, in the method `Device::process`. Seems OpenGL doesn't support an operation, could you give us your OpenGL version?

cc @kvark @msiglreith

**Comment by kvark** [comment]
I've actually seen that too yesterday, it's our (upstream) gfx-rs bug, but I haven't gotten to look into it or file it officially.

**Comment by torkleyy** [comment]
But this looks just like the usual clear command, so what could possibly go wrong here?

**Comment by Hittherhod** [comment]
@torkleyy OpenGL is 4.1, OpenCL is 1.2

**Comment by kvark** [comment]
@torkleyy something has died already and we are still trying to clear the buffer, perhaps

**Comment by Hittherhod** [comment]
@kvark The output says InvalidFrameBufferOperation, but from [http://docs.gl/gl4/glClearBuffer](url) it looks like that can only be output when calling the glClearNamedFrameBuffer* APIs. Those are only supported in OpenGL 4.5 and higher and my machine is using OpenGL 4.5. I'm not too familiar with the workings of gfx (I'm using amethyst to try and learn these things), but could it be a bug with detecting the version of OpenGL?

**Comment by Rhuagh** [comment]
I'm guessing it's a race condition between clearing up gl stuff (which is likely tied to the window), and the renderer running one last iteration before handing off control to the state for closing.

**Comment by Hittherhod** [comment]
Just an update: I found this bug when running even the gfx examples (specifically the mipmap example), so this might not actually be a problem with amethyst (https://github.com/gfx-rs/gfx/issues/1757)

**Comment by Xaeroxe** [comment]
Downgraded to priority normal, this doesn't impact gameplay it mostly just causes the program to take an unexpected path to exit.  Crashing right as the game is closed is probably the least harmful time to crash.

**Comment by francesca64** [comment]
Not only is this not a problem in Amethyst, but it's also not a problem in gfx.

It's fixed in tomaka/winit#476. Prior to that, `Closed` on macOS (and Windows, along with X11 as of 0.11.3) meant that the window was *already* closed, which made it very easy to exist in an invalid state.

**Comment by Xaeroxe** [comment]
@francesca64

Nice job!!  Thank you!  We'll close this one we've upgraded our winit

**Comment by jojolepro** [comment]
Did you notice the referenced PR is still opened?

**Comment by Xaeroxe** [comment]
Yup, we'll make sure to watch its progress prior to closing.

**Comment by Xaeroxe** [comment]
Fixed in winit 0.13, just waiting for 0.13 to be released and integrated into Amethyst.
